add support for indexing to meshgrid

dagitses opened the pull request:
This is step 3/7 of #50276. It only adds support for the argument but doesn't implement new indexing modes yet.

Test Plan:
Verified this is not FC breaking by adding logging to both meshgrid
overloads and then called meshgrid twice:

`meshgrid(*tensors)`
  and
`meshgrid(*tensors, indexing='ij')`

This confirmed that the former signature triggered the original native
function and the latter signature triggered the new native function.

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2133,8 +2133,22 @@ std::vector<Tensor> unbind(const Tensor& self, Dimname dim) {
+                    "indexing argument to torch.meshgrid.");
+    indexing = "ij";
+  } else {
+    TORCH_CHECK(*indexing == "ij", "Unsupported indexing: ", *indexing);
```
For these error messages start with the name of the operation and let people know what the correct indexing mode/s is/are. For example:

"torch.meshgrid: only 'ij' indexing is supported at this time"

dagitses replied:
should it still include the offending value?

mruberry replied:
That'd be nice

dagitses replied:
done.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5372,6 +5372,11 @@
 
 - func: meshgrid(Tensor[] tensors) -> Tensor[]
 
+# TODO: Two weeks after this lands, combine these two overloads,
```
Nice comment

dagitses replied:
😉

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5372,6 +5372,11 @@
+# TODO: Two weeks after this lands, combine these two overloads,
+#       making "indexing" optional. These are temporarily distinct for
+#       forward-compatibility reasons.
+- func: meshgrid.indexing(Tensor[] tensors, *, str? indexing) -> Tensor[]
```
Which of the two above signatures gets matched when calling `meshgrid(t)`?

Does that matching rely on the way PyTorch's dispatch system orders the signatures?

dagitses replied:
No, I think it 100% relies on the torch.functional() implementation. Does this impact the FC strategy? Do I need to change torch.functional() to continue delegating to the legacy implementation for the two week FC period?

mruberry replied:
I think so. See comment below

dagitses replied:
resolving this, let's take the discussion to the other comment thread.

mruberry commented on torch/functional.py:
```diff
@@ -386,16 +390,16 @@ def meshgrid(*tensors):
 
         """
-        return _meshgrid(*tensors)
+        return _meshgrid(*tensors, indexing=indexing)
```
If this is going to warn people to pass `indexing`, which seems very reasonable, then this should probably update the above example to always including the `indexing` kwarg and demonstrate our recommended practice

mruberry replied:
fyi not all the meshgrid uses in the above examples were updated to use "indexing"

dagitses replied:
True! That is intentional. I don't use it in the one where we plot the output. Most of the time in a plot you will want "xy" indexing, and eventually that becomes the default. I don't want to have a temporary example where we have `indexing='ij'`, since we don't really want anyone copying that.

I think when we add the option (i.e. next PR) to have `indexing='xy'` it makes sense to update that example.

mruberry commented on torch/functional.py:
```diff
@@ -386,16 +390,16 @@ def meshgrid(*tensors):
         # the old interface of passing the operands as one list argument
         tensors = tensors[0]  # type: ignore[assignment]
-    return _VF.meshgrid(tensors)  # type: ignore[attr-defined]
+    return _VF.meshgrid(tensors, indexing=indexing)  # type: ignore[attr-defined]
```
When someone calls meshgrid won't they call the functional and always call into the new signature from it using this approach?

dagitses replied:
Yes, per above comment, do we need to ensure that calls without indexing still go to the old version?

mruberry replied:
Yes, because otherwise programs that just call meshgrid will hit an FC issue and won't be able to run on older versions of PyTorch (even versions of PyTorch that are just hours old!)

dagitses replied:
is this what you had in mind as a fix?

mruberry commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2907,7 +2907,12 @@ def baddbmm(g, self, batch1, batch2, beta, alpha):
 
-def meshgrid(g, tensor_list):
+@parse_args('v', 's')
+def meshgrid(g, tensor_list, indexing: Optional[str] = None):
```
Nice ONNX update

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3987,11 +3987,12 @@ def sample_inputs(op_info: OpInfo, device: torch.device, dtype: torch.dtype,
 
         sample_inputs = []
-        for shapes in test_cases:
+        for shapes, indexing in itertools.product(test_cases, {'ij'}):
```
Nice use of product

dagitses replied:
😉

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6594,9 +6595,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
-           # Our implementation corresponds to "ij" indexing for
-           # numpy.meshgrid, but its default value is "xy".
-           ref=lambda *tensors: np.meshgrid(*tensors, indexing='ij'),
+           ref=np.meshgrid,
```
Cool

dagitses replied:
yeah, this change made me happy.

mruberry reviewed (commented):
Hey @dagitses! A few questions inline for your review. Also this needs a PR summary explaining what it does and, briefly, why. Looking forward to hearing your thoughts!

dagitses reviewed (commented):
updated the meshgrid tests now. I wanted to use `subTest` but the simple addition of `with self.subTest(indexing=None):` with no other changes led to the following failure that I don't understand:

```
======================================================================                                                                  
ERROR: test_meshgrid_meta (__main__.TestTensorCreationMETA) (indexing=None)                                                             
----------------------------------------------------------------------                                                                  
Traceback (most recent call last):                                                                                                      
  File "test/test_tensor_creation_ops.py", line 1381, in test_meshgrid                                                                  
    self.assertTrue(grid_a.equal(expected_grid_a))                                                                                      
NotImplementedError: Could not run 'aten::equal' with arguments from the 'Meta' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee usi
ng PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::equal' is only available for these backends: [CPU, QuantizedCPU, BackendSelect, Python, Named, Conjugate, Negative, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, Au
togradXLA, AutogradLazy, AutogradXPU, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, UNKNOWN_TENSOR_TYPE_ID, Autocast, Batched, VmapMode].
                                                                                                                                        
CPU: registered at aten/src/ATen/RegisterCPU.cpp:21984 [kernel]                                                                         
QuantizedCPU: registered at aten/src/ATen/RegisterQuantizedCPU.cpp:1060 [kernel]                                                        
BackendSelect: fallthrough registered at ../aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]                     
Python: registered at ../aten/src/ATen/core/PythonFallbackKernel.cpp:38 [backend fallback]                                              
Named: fallthrough registered at ../aten/src/ATen/core/NamedRegistrations.cpp:11 [kernel]                                               
Conjugate: registered at ../aten/src/ATen/ConjugateFallback.cpp:26 [backend fallback]                                                   
Negative: registered at ../aten/src/ATen/native/NegateFallback.cpp:26 [backend fallback]                                                
ADInplaceOrView: fallthrough registered at ../aten/src/ATen/core/VariableFallbackKernel.cpp:64 [backend fallback]
AutogradOther: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]                                
AutogradCPU: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
AutogradCUDA: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]                                 
AutogradXLA: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
AutogradLazy: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
AutogradXPU: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
AutogradMLC: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]                                  
AutogradHPU: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
AutogradNestedTensor: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]                         
AutogradPrivateUse1: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]                          
AutogradPrivateUse2: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
AutogradPrivateUse3: registered at ../torch/csrc/autograd/generated/VariableType_3.cpp:12536 [autograd kernel]
Tracer: registered at ../torch/csrc/autograd/generated/TraceType_3.cpp:11354 [kernel]                                                   
UNKNOWN_TENSOR_TYPE_ID: fallthrough registered at ../aten/src/ATen/autocast_mode.cpp:444 [backend fallback]                             
Autocast: fallthrough registered at ../aten/src/ATen/autocast_mode.cpp:283 [backend fallback]                                           
Batched: registered at ../aten/src/ATen/BatchingRegistrations.cpp:1016 [backend fallback]
VmapMode: fallthrough registered at ../aten/src/ATen/VmapModeRegistrations.cpp:33 [backend fallback]                                    
```

Commit:
add support for indexing to meshgrid

Test Plan:
Verified this is not FC breaking by adding logging to both meshgrid
overloads and then called meshgrid twice:

`meshgrid(*tensors)`
  and
`meshgrid(*tensors, indexing='ij')`

This confirmed that the former signature triggered the original native
function and the latter signature triggered the new native function.

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1382,6 +1414,42 @@ def test_meshgrid(self, device):
         self.assertTrue(grid_b2.equal(expected_grid_b))
         self.assertTrue(grid_c2.equal(expected_grid_c))
 
+    def test_meshgrid_ij_indexing(self, device):
```
In the future it'll be interesting to compare against NumPy directly, too

dagitses replied:
we do that in the OpInfo via reference. Is there a different flavor of comparison we should do here?

mruberry replied:
SampleInputs are usually a small set of valid inputs because they're used to generate so many tests. A more specific "forward" test is a good way to exercise a wider and more exhaustive range of inputs

dagitses replied:
That makes sense: we only want this heavy volume of inputs for the reference test.

Do you have an example test I can look at so I can gauge how comprehensive to be? I'll write the test in the meantime.

mruberry replied:
It should exhaustively test the operator's code paths and stress edge case behavior. The exact number of cases is a little difficult to prescribe in advance. For some references here's the test cases for nextafter:

https://github.com/pytorch/pytorch/blob/1c2b5e59ae53ef3042ad1b5cf9aa275391c92971/test/test_binary_ufuncs.py#L1875

and here's for einsum:

https://github.com/pytorch/pytorch/blob/d37636901ed1c65c1f8b68e36e37e59eb503c554/test/test_linalg.py#L4493

dagitses replied:
implemented!

mruberry reviewed (approved):
Thank you for your patience on this PR, @dagitses. This looks good. See just one small documentation suggestion inline.

Given how close this review is to the branch cut (sorry about that!) I think we should hold it until after 1.10 is cut.

dagitses commented:
> Thank you for your patience on this PR, @dagitses. This looks good. See just one small documentation suggestion inline.
> 
> Given how close this review is to the branch cut (sorry about that!) I think we should hold it until after 1.10 is cut.

No problem. There's no need to rush this, so if it's too risky, it can wait till after the release.

I will update to `'xy'` indexing in the plot example in the following PR.

dagitses commented:
> > Thank you for your patience on this PR, @dagitses. This looks good. See just one small documentation suggestion inline.
> > Given how close this review is to the branch cut (sorry about that!) I think we should hold it until after 1.10 is cut.
> 
> No problem. There's no need to rush this, so if it's too risky, it can wait till after the release.
> 
> I will update to `'xy'` indexing in the plot example in the following PR.

that doc change is pushed now in #62724

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1382,6 +1414,82 @@ def test_meshgrid(self, device):
+            ({}, {'indexing': 'ij'}),
+            # "ij" is implemented identically in both.
+            ({'indexing': 'ij'}, {'indexing': 'ij'}),
+            # TODO Test "xy" when it is supported in PyTorch.
```
I would move this test to the PR where xy indexing is added, too, and just write the test always specifying the indexing for simplicity

dagitses replied:
Fair enough. Do you mind if I leave this since there is some value in testing the defaults as well? It will only live on for another six months.

dagitses replied:
thanks!

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1382,6 +1414,82 @@ def test_meshgrid(self, device):
+    def test_meshgrid_vs_numpy(self, device):
+        # Possible shapes to the random tensors. Note that [] produces
+        # a 0D tensor.
+        shapes = [[], [1], [2], [3], [5]]
```
I suggest just handwriting cases with special focus on weird edge cases

dagitses replied:
yeah, i agree that is better. do you care if the shapes are manually curated but we still do random values (really shouldn't matter for this op)? Or should I be fully determinstic with the tensors too?

mruberry replied:
Manual shapes with random values sounds great (using randn or make_tensor) sounds great

dagitses replied:
great done.
